File: minto/5주차_DP/1932.js
const input = require('fs').readFileSync('/dev/stdin').toString().trim().split('\n');

const n = input.shift();

const pyramid = input.map((item) => item.split(' '));

for (let i = 1; i < pyramid.length; i++) {
  if (i === 1) {
    pyramid[1][0] = +pyramid[1][0] + +pyramid[0];
    pyramid[1][1] = +pyramid[1][1] + +pyramid[0];
  } else {
    pyramid[i][0] = +pyramid[i][0] + pyramid[i - 1][0];
    pyramid[i][i] = +pyramid[i][i] + pyramid[i - 1][i - 1];

    for (let j = 1; j < pyramid[i].length - 1; j++) {
      pyramid[i][j] = +pyramid[i][j] + Math.max(pyramid[i - 1][j - 1], pyramid[i - 1][j]);
    }
  }
}

console.log(Math.max(...pyramid[n - 1]));

/**
 * 아래 값들을 이용해서 최대값을 구면서 새로운 배열을 만드는 방식으로 푼 DP 문제이다.
 */

File: ujin/5주차_DP/1932.py
n = int(input())
tri = [list(map(int, input().split())) for _ in range(n)]

dp = [[0] * n for _ in range(n)]
dp[0][0] = tri[0][0]

k = 2
for i in range(1, n):
    for j in range(k):
        if j == 0:
            dp[i][j] = dp[i - 1][j] + tri[i][j]
        elif i == j:
            dp[i][j] = dp[i - 1][j - 1] + tri[i][j]
        else:
            dp[i][j] = max(dp[i - 1][j - 1], dp[i - 1][j]) + tri[i][j]
    k += 1

print(max(dp[n-1]))

'''
각 자리에 올 수 있는 가장 큰 값을 dp에 저장하고, 마지막 줄의 숫자 중 최댓값을 출력하도록 했다.
'''
